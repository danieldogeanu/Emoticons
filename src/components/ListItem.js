import React from 'react';
import CopyButton from './CopyButton';
import Emoticon from '../elements/Emoticon';
import Name from '../elements/Name';
import Copy from '../elements/Copy';
import CopyHint from '../elements/CopyHint';
import '../styles/components/ListItem.scss';

function ListItem(props) {
	const {data, type, style} = props;
	const {char, name} = data;

	switch (type) {
		case 'desktop':
			return (
				<li className="DesktopListItem" data-testid="DesktopListItem" style={style}>
					<Emoticon char={char} name={name} />
					<Name name={name} />
					<Copy>
						<CopyButton data={char} />
					</Copy>
				</li>
			);
		case 'mobile':
			return (
				<li className="MobileListItem" data-testid="MobileListItem" style={style}>
					<CopyButton data={char} className="mobileCard ripple">
						<Name name={name} />
						<div className="bottomBar">
							<Emoticon char={char} name={name} />
							<CopyHint text="Tap to Copy" />
						</div>
					</CopyButton>
				</li>
			);
		default:
			return false;
	}
}

export default ListItem;
