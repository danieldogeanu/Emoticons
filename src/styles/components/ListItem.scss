@import '../includes';

.DesktopListItem {
	display: flex;
	height: rem(50);
	padding: rem(6) 0;
	transition: background-color 0.3s ease;
	&:hover { background-color: color-scheme(item, bg-hover); }

	.Emoticon {
		width: rem(80);
	}

	.Name {
		display: flex;
		align-items: center;
		flex-grow: 1;
	}

	.Copy {
		display: flex;
		justify-content: center;
		align-content: center;
		width: rem(80);
	}
}

.Emoticon {
	display: flex;
	justify-content: center;
	align-content: center;
	font-size: rem(22);
	user-select: none;
	color: color-scheme(emoticon);
	@include breakpoint(min, rem(481)) {
		font-size: rem(26);
	}
}

.Name {
	font-size: rem(14);
	font-weight: weight(light);
	font-family: font(open-sans);
	text-transform: capitalize;
	color: color-scheme(text);
	@include breakpoint(min, rem(301)) { font-size: rem(16); }
	@include breakpoint(min, rem(641)) { font-size: rem(18); }
	@include breakpoint(min, rem(851)) { font-size: rem(20); }
}