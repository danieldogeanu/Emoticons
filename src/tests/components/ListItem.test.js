import React from 'react';
import {render} from '@testing-library/react';
import ListItem from '../../components/ListItem';

const testEmoticon = {
	"codes": "1F604",
	"char": "ðŸ˜„",
	"name": "grinning face with smiling eyes",
	"category": "Smileys & Emotion (face-smiling)",
	"group": "Smileys & Emotion",
	"subgroup": "face-smiling"
};

describe('ListItem Component', () => {

	it('renders desktop list item properly', () => {
		const {getByTestId, container} = render(<ListItem type="desktop" data={testEmoticon} />);
		const renderedItem = getByTestId('DesktopListItem');
		const renderedEmoticon = container.querySelector('.Emoticon');
		const renderedName = container.querySelector('.Name');
		const renderedCopyBtn = container.querySelector('.CopyButton');

		expect(renderedItem).toBeInTheDocument();
		expect(renderedItem).toHaveClass('DesktopListItem');
		expect(renderedItem).toContainElement(renderedEmoticon);
		expect(renderedItem).toContainElement(renderedName);
		expect(renderedItem).toContainElement(renderedCopyBtn);
		expect(renderedItem).toMatchSnapshot();

		expect(renderedEmoticon).toHaveTextContent(testEmoticon.char);
		expect(renderedEmoticon).toHaveAttribute('aria-label', testEmoticon.name);
		expect(renderedName).toHaveTextContent(testEmoticon.name);
		expect(renderedCopyBtn).toHaveAttribute('data-clipboard-text', testEmoticon.char);
	});

	it('renders mobile list item properly', () => {
		const {getByTestId, container} = render(<ListItem type="mobile" data={testEmoticon} />);
		const renderedItem = getByTestId('MobileListItem');
		const renderedCopyBtn = container.querySelector('.CopyButton');
		const renderedName = container.querySelector('.Name');
		const renderedEmoticon = container.querySelector('.Emoticon');

		expect(renderedItem).toBeInTheDocument();
		expect(renderedItem).toHaveClass('MobileListItem');
		expect(renderedItem).toContainElement(renderedCopyBtn);
		expect(renderedItem).toContainElement(renderedName);
		expect(renderedItem).toContainElement(renderedEmoticon);
		expect(renderedItem).toMatchSnapshot();

		expect(renderedCopyBtn).toHaveAttribute('data-clipboard-text', testEmoticon.char);
		expect(renderedCopyBtn).toHaveClass('mobileCard', 'ripple');
		expect(renderedCopyBtn).toContainElement(renderedName);
		expect(renderedCopyBtn).toContainElement(renderedEmoticon);

		expect(renderedName).toHaveTextContent(testEmoticon.name);
		expect(renderedEmoticon).toHaveTextContent(testEmoticon.char);
		expect(renderedEmoticon).toHaveAttribute('aria-label', testEmoticon.name);
	});

});
