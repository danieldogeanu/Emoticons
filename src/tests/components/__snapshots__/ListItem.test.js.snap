// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem Component renders desktop list item properly 1`] = `
<li
  class="DesktopListItem"
  data-testid="DesktopListItem"
>
  <span
    aria-label="grinning face with smiling eyes"
    class="Emoticon"
    role="img"
  >
    ðŸ˜„
  </span>
  <span
    class="Name"
  >
    grinning face with smiling eyes
  </span>
  <span
    class="Copy"
    title="Copy the emoticon and paste it where you need it."
  >
    <button
      class="CopyButton undefined"
      data-clipboard-text="ðŸ˜„"
      title=""
      type="button"
    >
      Copy
    </button>
  </span>
</li>
`;

exports[`ListItem Component renders mobile list item properly 1`] = `
<li
  class="MobileListItem"
  data-testid="MobileListItem"
>
  <button
    class="CopyButton mobileCard ripple"
    data-clipboard-text="ðŸ˜„"
    title=""
    type="button"
  >
    <span
      class="Name"
    >
      grinning face with smiling eyes
    </span>
    <div
      class="bottomBar"
    >
      <span
        aria-label="grinning face with smiling eyes"
        class="Emoticon"
        role="img"
      >
        ðŸ˜„
      </span>
      <span
        class="CopyHint"
      >
        Tap to Copy
      </span>
    </div>
  </button>
</li>
`;
